### ISSUE TITLE
Resumable downloading required

### ISSUE THREAD

**Issue by ZuYuanZhou** [SDWebImage/SDWebImage#867]
Is it possible to add resumable downloading feature in coming version?

I find that if I want to download big size images in UICollectionViewCell, when the cell is reused, the previous image downloading will be cancelled. That's to say, if I scroll back, the image need to re-download again even it has be downloaded 99%. If I use a progress view in the UICollectionViewCell to show the downloading progress, it will from 99% back to 0.

**Comment by skyline75489** [comment]
Now that we've changed to `NSURLSession`, is this implemented ? If not, will this be implemented any time soon? @bpoplauschi 

I'm willing to implement and PR this if help is needed.

**Comment by bpoplauschi** [comment]
`NSURLSession` has support for suspend/resume. I think it's important (in most cases) that we cancel the old downloads when trying to load new images (i.e. when scrolling in a table) to free up the network resources. The iOS system has a maximum number of active requests it can support at a time (at least from the same host). Even if you queue more requests, they will be inactive.
The proposal here was to suspend them instead of cancel and try to resume if we come back. It might work, not sure. @skyline75489 feel free to try to implement, but I'm not convinced this is a feature that needs to be added to the library itself rather than being built in different apps. 

Other opinions?

**Comment by skyline75489** [comment]
I think that "suspending" isn't ideal for the reason you talked about. We need to free up the network resources. However, it would make sense if we use `cancelByProducingResumeData` to temporaily save the data downloaded. Later on we can use `downloadTaskWithResumeData` to continue the download.

**Comment by bpoplauschi** [comment]
I was not aware of the `cancelByProducingResumeData:` or `downloadTaskWithResumeData:`. Those sound interesting.

**Comment by skyline75489** [comment]
This weekend I'll try to figure out a demo, maybe even a PR, to demostrate this feature.

**Comment by skyline75489** [comment]
I did a little research tonight. Although `NSURLSessionDownloadTask`'s methods looks sweet and all, it has a serious flaw (perhaps design in Apple's perspective), that is saving the data to file directly after downloading. While downloading, you can only get the received and expected byte count, but not the actual data. This violates so many things that `SDWebImageDownloaderOperation` are doing right now, such as progressively showing the image and compressing the image if needed.

This is not as easy as I thought it should be. Man, has SDWebImage changed since 3.8.x. AFAIK, there seems to be no other way to implement this (without breaking other parts of the library) other than manually handling the resuming, which is basically writing a `NSURLSessionDownloadTask`. Hopefully someday me or some other guys would come up with a better idea.

**Comment by bpoplauschi** [comment]
@skyline75489 I see. It sounds indeed complicated and the goal for now is to release and stabilise the current 4.0.0. Let's keep this ticket open and revisit later. Thanks for investigating.

**Comment by novinfard** [comment]
+1
I couldn't find any way to do it manually, but it is necessary for it's efficiency. Everyone should look on 4.x as mature release and not having this feature is a disappointing issue.

**Comment by dreampiggy** [comment]
emmm...Why can we have a priority queue and limit the size of concurrent download operation. When the cell reuse, we do not acutally cancel the download operation, just cancel the process to `set the image`. More clearly, is the `SDWebImageDownloadToken`'s completionBlock.

If we seperate the download completion from download itself, we can keep the download go on without call the completion block. Just use another private block or context to save the data to the cache(Maybe not memory cache, disk cache only) and when next time the same URL request came, we can retrive the response directlly from the cache to avoid this issue. And this can also solve the issue when reuse cell the progress state is cleared(The progress state is kept in private block or context).

**Comment by dreampiggy** [comment]
In 5.x, we now have that `private context` which can be passed from top level module to the bottom and back called SDWebImageContext, so maybe it’s possibe to implement this feature from the architecture.

We can flag a valie in context, if the value set to disable cancel, when the cell reuse and trigger a `cancel` for download operation, the network will not be cancelled and continue, process normal logic unless including writing to cache, but not trigger a final set image block and completion block. Next time you request the same URL, the cache will be hited and return a image immediatelly.

However, because this change the pipeline of we process download, cache, and manage view category, the solution may be considered carefully. If I have some time I can try a naive approach and let other contributors review. I thin it’s useful to keep this feature request continue.

**Comment by dreampiggy** [comment]
Changing the task from the current `DataTask` to `DownloadTask` is far more complicated. Because we have progressive download, which need the actual partial bytes in memory to decode bitmap. However, `DownloadTask` is file path based API (only tell you the progress but not partial bytes data). So if we want to use resume data, it’s hard.

My solution is to not change the download logic, keep using DataTask, but do a extra management for manager-view category-downloader. These 3 module need to communicate about `whether current image request is un-cancellable`. But this naive approach need to consider how to solve the possible explosive `non-cancellable` task fill the network request queue.

**Comment by zhongwuzw** [comment]
@dreampiggy Make request `un-cancellable` may leads to more issues, the memory(may leads to more OOM), cpu, I/O, network load (keep the socket connection, needs more time to complete and lose the management of priority) .
If we try to implement resumable of download, I think we can just simulate the resumable process of `HTTP`, check `Accept-Ranges` header , if server support resumable, before we cancel the request, we can save the response `Etag` 、`Last-Modified` and received data to disk, next time when we load the same url, we'd use these info to try to get rest of data if server not modify the data.

**Comment by dreampiggy** [comment]
@zhongwuzw What you say it’s just DownloadTask. Apple already done for you. It’s suspendable and resumble.

But the reason why we choose DataTask contains many aspects. Chaning this is also evil. My idea is directly aim to solve the user’s request: “Don’t cancel download when cell resuing”. Which does not menas “suspend when cell disappear, resume when cell appear for reuse”. These two apects are really different.

**Comment by zhongwuzw** [comment]
@dreampiggy `Apple` can do it by using `NSURLSessionDownloadTask`, but we can't use it, because we need the progressive data.

**Comment by zhongwuzw** [comment]
@dreampiggy 
> But the reason why we choose DataTask contains many aspects. Chaning this is also evil. My idea is directly aim to solve the user’s request: “Don’t cancel download when cell resuing”. Which does not menas “suspend when cell disappear, resume when cell appear for reuse”. These two apects are really different.
I get your point, and I said the issues it may leads 😂  :
> Make request un-cancellable may leads to more issues, the memory(may leads to more OOM), cpu, I/O, network load (keep the socket connection, needs more time to complete and lose the management of priority) .

**Comment by zhongwuzw** [comment]
I think we need to keep in mind is if user cancel, we need cancel everything, if user start loading again, we try to decrease unnecessary operations.

**Comment by zhongwuzw** [comment]
If user don't want to cancel, I think he can use `ImageDownloader` directory, decouple the `UIImageView` in cell and image.

**Comment by dreampiggy** [comment]
Yes. My idea is quite simple: Since what they want, is `when I trigger another `sd_setImageWithURL:`, don't cancel the previous download network request, instead, keep it downloading, but don't do a actual image refresh and completionBlock`.

So I think this detailed feature request we can do. User who want this behavior should provide a extra options like `SDWebImageAvoidCancelCurrentLoad`. Then, we can do our logic to solve this, such as remove the previous `completionBlock`associated to previous URL (actually the `SDWebImageDownloadToken` we use), avoid trigger a cancel on `SDWebImageCombinedOperation`. So I think this feature should related to both `View Category`, `Downloader`, `Manager` 3 modules. However, previouslly in 4.x, we have no communication way between different module (we have only `enum`). But now we can use context for communication. So it's why I think maybe this feature can be done with current architecture.

**Comment by zhongwuzw** [comment]
Many things seems can't solve, I just list two, images is a really memory terminator, very easy to cause OOM, and if users use `HTTP1.1`, previous requests would block the new request, so when scroll the `tableView` to the new cell, very long time would be to wait for completion, mostly it would timeout IMO. We can add an option easily, but I think it may unusable.

**Comment by dreampiggy** [comment]
`previous requests would block the new request` this one ? What's your meaning ? If you means the `HTTPMaximumConnectionsPerHost`, I remember you told that it's not a case we as a framework should solve or assume.

**Comment by zhongwuzw** [comment]
Head-of-Line blocking.
<img width="1030" alt="image" src="https://user-images.githubusercontent.com/5061845/47606155-98241b80-da42-11e8-8ccf-bfdba698e02c.png">
